### ISSUE TITLE
Confusion Matrix

### ISSUE THREAD

**Issue by QuarTerll** [ultralytics/yolov5#8779]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

When I run val.py, I got a valition results from the terminal and a picture of confusion matrix.

In my opinion, I thought the precison computed from confusion matrix .jpg should be same as the results from the terminal.

But it is not.

Should it be the same?

**Comment by glenn-jocher** [comment]
@QuarTerll sorry for the confusion. Confusion matrix uses default inference thresholds of 0.25 conf 0.45 iou while console results are output at max F1 confidence which varies per class, which will return the highest P and R balance. You can update console output here:
https://github.com/ultralytics/yolov5/blob/567397d67ae173fb82e06672a763cc28c5cfeb2b/utils/metrics.py#L89-L94

**Comment by QuarTerll** [comment]
> @QuarTerll sorry for the confusion. Confusion matrix uses default inference thresholds of 0.25 conf 0.45 iou while console results are output at max F1 confidence which varies per class, which will return the highest P and R balance. You can update console output here:
> 
> https://github.com/ultralytics/yolov5/blob/567397d67ae173fb82e06672a763cc28c5cfeb2b/utils/metrics.py#L89-L94

@glenn-jocher Thanks for reply.

But when I check for this, I found another question.

(I have not change the code about your reply. So it is the original code.)
When I finished training, I got a pic of confusion matrix and a validation results from the terminal on the best.pt

For now, I just use the val.py to reproduce it with threshold 0.5 and 0.5 on the best.pt
But I get the different results. 

I check the code. I found that, while training, in val.py the theshold is 0.001 and 0.6.
So I use 0.001 and 0.6 to reproduce it again.

Still get the different results from the training outputs.

What happend here?
Should it be the same?


Here is how I run the val.py
`python val.py --verbose --project /root/runs/val --name test4any --weights /root/runs/train/*some_path*/weights/best.pt --data *data_config*.yaml --conf-thres 0.001 --iou-thres 0.6 --task val --device 0,1,2,3 `


Below is some logs and outputs.

validation ouputs:
![image](https://user-images.githubusercontent.com/49419815/182305369-02d42615-cd8b-4dbe-aac2-11da037db7cc.png)

![image](https://user-images.githubusercontent.com/49419815/182305503-5a17fe4e-63d3-455b-b46f-b217b26bf636.png)


Confusion Matrix:
End of the training.
![image](https://user-images.githubusercontent.com/49419815/182302125-ed77b668-f6a5-4ae2-b5b5-da759e523f1e.png)
Reproduce with val.py
![image](https://user-images.githubusercontent.com/49419815/182302840-3fa28ce3-5192-4d1f-8f3c-c7487e545c1c.png)

**Comment by QuarTerll** [comment]
@glenn-jocher BTW, why using 0.001 to validate while training ?
If so, what thresholds should I use in detect.py?
Should I use 0.001 as well?

In normal case, I thought it should be 0.5.

**Comment by glenn-jocher** [comment]
@QuarTerll 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!

**Comment by QuarTerll** [comment]
@glenn-jocher Thanks for your detailed reply, which really helped me a lot of about my confusions.

So, We get almost all detections while training and validating in training for the best results.

As for detect.py and YOLOv5 PyTorch Hub Inference, it is designed for inference of different data source.
Then, why the threshold is 0.25 and 0.45 in inference. (cus it is the best for coco in yolov5?)

As you said, the results given by val.py is the outputs on best F1. So Is 0.25 and 0.45 the best threshold in best F1?

If so, the outputs of terminal only give metrics but not relevant thresholds for each class. 
**Then how can I get the best threshold for my custom dataset? 
In other words, how can I get the confidence threshold for detect.py or YOLOv5 PyTorch Hub Inference?**

**Comment by QuarTerll** [comment]
BTW, before today, I just thought the two thresholds in val.py is same as them in detect.py. LOL
But in fact, there are some small differences. hmmmmm

For other beginners same as me......
The inputs thresholds in val.py defines what the model output but not directly relevant to the results of metrics from the terminal. 
The showed metrics is being computed by the thresholds of the best F1 based on the model outputs filtered by the inputs thresholds in val.py

**Comment by glenn-jocher** [comment]
@QuarTerll 0.25 conf 0.45 IoU is standard detections thresholds across tools/platforms.

val.py goes a bit further and examines the best F1 for each class. You can see this in the F1 cuve png that is produced by val.py.

It's up to the domain expert to determine suitable deployment settings to meet your priorities, i.e. higher P or higher R.

**Comment by glenn-jocher** [comment]
@QuarTerll PyTorch Hub thresholds are here, you can change them to whatever you want, i.e. `model.conf = 0.25`
https://github.com/ultralytics/yolov5/blob/4d8d84b0ea7147aca64e7c38ce1bdb5fbb9c5a53/models/common.py#L547-L556

**Comment by QuarTerll** [comment]
@glenn-jocher gocha. Thanks for you patience. Now I have more understanding about yolov5. :)

**Comment by QuarTerll** [comment]
> @QuarTerll 0.25 conf 0.45 IoU is standard detections thresholds across tools/platforms.
> 
> val.py goes a bit further and examines the best F1 for each class. You can see this in the F1 cuve png that is produced by val.py.
> 
> It's up to the domain expert to determine suitable deployment settings to meet your priorities, i.e. higher P or higher R.



As for the next step, I am going to try to add some codes on the part of terminal outputs for the best thresholds showing. 
This maybe help to understand this process and give advices to define the thresholds for detect.py for each class.

I will let you know if I finished this. 
(Maybe next week hmmmm
(This maybe not easy for me :(

Something like below

![image](https://user-images.githubusercontent.com/49419815/182751751-9591f3fc-18ff-4687-b834-bd41648eaad5.png)
